dill/_dill.py

```diff
@@ -1548,51 +1655,57 @@ def save_classmethod(pickler, obj):
-           #        del globs[key]
-            # ABORT: if self-references, use _recurse=False
-            if id(obj) in stack: # or obj in globs.values():
-                globs = obj.__globals__ if PY3 else obj.func_globals
         else:
             globs = obj.__globals__ if PY3 else obj.func_globals
-        _byref = getattr(pickler, '_byref', None)
-        _recurse = getattr(pickler, '_recurse', None)
-        _memo = (id(obj) in stack) and (_recurse is not None)
-       #print("stack: %s + '%s'" % (set(hex(i) for i in stack),hex(id(obj))))
-        stack[id(obj)] = len(stack), obj
-        if PY3:
-            #NOTE: workaround for 'super' (see issue #75)
-            _super = ('super' in getattr(obj.__code__,'co_names',())) and (_byref is not None)
-            if _super: pickler._byref = True
-            if _memo: pickler._recurse = False
-            fkwdefaults = getattr(obj, '__kwdefaults__', None)
-            pickler.save_reduce(_create_function, (obj.__code__,
-                                globs, obj.__name__,
-                                obj.__defaults__, obj.__closure__,
-                                obj.__dict__, fkwdefaults), obj=obj)
+        if _memo:
+            pickler._recurse = False
+            if PY3:
+                #NOTE: workaround for 'super' (see issue #75)
+                _super = ('super' in getattr(obj.__code__,'co_names',())) and (_byref is not None)
+                fkwdefaults = getattr(obj, '__kwdefaults__', None)
+                pickler.save_reduce(_create_function, (obj.__code__,
+                                    globs, obj.__name__,
+                                    obj.__defaults__, obj.__closure__,
+                                    obj.__dict__, fkwdefaults), obj=obj)
+            else:
+                _super = ('super' in getattr(obj.func_code,'co_names',())) and (_byref is not None) and getattr(pickler, '_recurse', False)
+                if _super: pickler._byref = True
+                pickler.save_reduce(_create_function, (obj.func_code,
+                                    globs, obj.func_name,
+                                    obj.func_defaults, obj.func_closure,
+                                    obj.__dict__), obj=obj)
+                if _super: pickler._byref = _byref
+            pickler._recurse = _recurse
         else:
-            _super = ('super' in getattr(obj.func_code,'co_names',())) and (_byref is not None) and getattr(pickler, '_recurse', False)
-            if _super: pickler._byref = True
-            if _memo: pickler._recurse = False
-            pickler.save_reduce(_create_function, (obj.func_code,
-                                globs, obj.func_name,
-                                obj.func_defaults, obj.func_closure,
-                                obj.__dict__), obj=obj)
-        if _super: pickler._byref = _byref
-        if _memo: pickler._recurse = _recurse
+            if PY3:
```

Comment:
having a cut and paste x4 is a bit unwieldy.  I'm wondering if it'd be better to have a if block that defines a function as a direct call to `save_reduce` or a `_save_with_postproc`'d call.  Then we could avoid duplication of the code.

Reply:
resolved